### ISSUE TITLE
Ensure non-printable characters aren't reported for color-contrast

### ISSUE THREAD

**Issue by marcysutton** [dequelabs/axe-core#315]
Our color contrast rule is reporting incomplete for non-printable characters, such as:

```html
<span>&rlm;</span>
``` 

As a result, it's really difficult to debug them in the new Needs Review UI (to be released) because they have no visible text content. The related node ends up as:

```html
<span></span>
```

We should exclude these elements in `color-contrast-matches.js`, along with any other non-printable characters. Do some research to build up a list of candidates in addition to `&rlm;`.

A live example can be found in the Notifications page of Twitter.com.

**Comment by WilcoFiers** [comment]
This might be a good starting point: http://www.fileformat.info/info/unicode/category/index.htm

**Comment by straker** [comment]
So I dug into this. JavaScript doesn't really understand unicode categories, so we couldn't use a Regex to just test for certain categories. There's a proposal to implement [category aware regex syntax](https://github.com/tc39/proposal-regexp-unicode-property-escapes). It has made it to stage 4 and will be implemented in ES2018, but that wouldn't help us with IE11. 

It does list 2 other libraries that fill in support for unicode: [XRegExp](https://github.com/slevithan/xregexp) and [Regenerate](https://github.com/mathiasbynens/regenerate). Regenerate works with IE6+, while XRegExp works with all ES5+ browsers and node. Both have their pros and cons which are mentioned in the proposal.

There's also the problem that what is "printable" in unicode is not always obvious. From a [stack overflow answer](https://stackoverflow.com/a/53450014/2124254):

> The more you learn about Unicode, the more you realise how unexpectedly diverse and unfathomably weird human writing systems are.
>
> In particular whether a particular "character" is printable is not always obvious.
> 
> Is a zero-width space printable? When is a hyphenation point printable? Are there characters whose printability depends on their position in a word or on what characters are adjacent to them? Is a combining-character always printable?

Another problem would be an element with just an emoji in it. An emojis color isn't determined by the style of the element, but axe still reports it as a color contrast error.

```html
<div style=" background: black; color: #3e3e3e; ">😋</div>
```

If we also want to exclude emoji from the color checker, we might run into a Sisyphus problem with a blacklist approach as the emoji library is constantly growing. 

We could try a whitelist approach. There's a library that contains a list of all unicode from the [L,M,N and Pc categories](https://github.com/paulmillr/unicode-categories/blob/master/index.js) and it's a really long list of unicode. It's also missing other printable characters from categories P and S, so we would need to expand it. However, testing every elements `textContent` on the page for the existence of at least one of these valid characters might prove to be quite slow.

We could also say that even if the element doesn't have a printable character, it poses a potential accessibility issue in that if the text were to change but the color remained the same, it would have a contrast issue.

**Comment by AutoSponge** [comment]
A strategy I've considered is using https://www.unicode.org/cldr/charts/latest/summary/root.html to warn against any extra-language characters. These regexes are usually much smaller and can be loaded dynamically to match the browser's pref lang and/or page lang. Also, they don't change (often?).

**Comment by AutoSponge** [comment]
Also, a technique used by github for emoji detection cloned text to canvas and performed pixel measurements. While computationally expensive (if done often), requires very little code.

**Comment by dylanb** [comment]
I think we should split out support for non-printable characters from emojis as they represent different problems (emojis could still have color contrast issues and probably fall under the new WCAG 2.1 S.C. 1.4.11 - I say "probably" because we have not made a definitive decision on this yet)

**Comment by AutoSponge** [comment]
If the regexp is unreliable (OS/font), we could use a simplified version of the canvas technique:

```js
var canvas = document.createElement('canvas');
canvas.width = 10;
canvas.height = 10;
var context = canvas.getContext('2d');
context.font = '10px System';  // match font?
context.textAlign = 'left';
context.textBaseline = 'top';
context.fillText('\u{0081}', 0, 0);
var { data } = context.getImageData(0, 0, 10, 10);
data.findIndex(n => n !== 0); // -1 means it didn't print anything
```

**Comment by AutoSponge** [comment]
```js
var canvas = document.createElement('canvas');
canvas.width = 10;
canvas.height = 10;
var context = canvas.getContext('2d');
context.font = '10px'; // maybe match font used?
context.textAlign = 'left';
context.textBaseline = 'top';
context.fillText('\u{001f}', 0, 0); // normally use elm.textContent if length === 1
var {data} = context.getImageData(0, 0, 10, 10);
data.findIndex(n => n !== 0); // -1 
```
This should reliably test any character.
